src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs

```diff
@@ -385,17 +387,27 @@ private Task ProcessNegotiate(HttpContext context, HttpConnectionOptions options
             // Connection ID metadata set.
             logScope.ConnectionId = connection.ConnectionId;
 
-            // Get the bytes for the connection id
-            var negotiateResponseBuffer = GetNegotiatePayload(connection.ConnectionId, context, options);
+            // Don't use thread static instance here because writer is used with async
+            var writer = new MemoryBufferWriter();
 
-            Log.NegotiationRequest(_logger);
+            try
+            {
+                // Get the bytes for the connection id
+                WriteNegotiatePayload(writer, connection.ConnectionId, context, options);
 
-            // Write it out to the response with the right content length
-            context.Response.ContentLength = negotiateResponseBuffer.Length;
-            return context.Response.Body.WriteAsync(negotiateResponseBuffer, 0, negotiateResponseBuffer.Length);
+                Log.NegotiationRequest(_logger);
+
+                // Write it out to the response with the right content length
+                context.Response.ContentLength = writer.Length;
+                return writer.CopyToAsync(context.Response.Body);
```

Comment:
You need to await here of the MemoryBufferWriter is going to be returned before the operation is complete *if* it goes async.

Reply:
Good catch